# Importing Dependencies

In [1]:
!uv add pandas pyarrow
import pandas as pd
import pyarrow.parquet as pq

Resolved 120 packages in 5ms
Audited 105 packages in 16ms


In [2]:
# Install SQL Alchemy so we can interact with PostgreSQL
!uv add sqlalchemy psycopg2-binary
from sqlalchemy import create_engine
engine = create_engine('postgresql://root:root@localhost:5450/homework')

# To view data being inserted into PostgreSQL
!uv add tqdm
from tqdm.auto import tqdm

Resolved 120 packages in 3ms
Audited 105 packages in 1ms
Resolved 120 packages in 3ms
Audited 105 packages in 1ms


# Reading Dataset

In [3]:
trips_parquet_file_loc = "./data/green_tripdata_2025-11.parquet"
zones_csv_file_loc = "./data/taxi_zone_lookup.csv"

In [4]:
df_trips = pd.read_parquet(trips_parquet_file_loc)
df_zones = pd.read_csv(zones_csv_file_loc)

In [5]:
df_trips.head()

,VendorID,lpep_pickup_datetime,lpep_dropoff_datetime,store_and_fwd_flag,RatecodeID,PULocationID,DOLocationID,passenger_count,trip_distance,fare_amount,...,mta_tax,tip_amount,tolls_amount,ehail_fee,improvement_surcharge,total_amount,payment_type,trip_type,congestion_surcharge,cbd_congestion_fee
0,2,2025-11-01 00:34:48,2025-11-01 00:41:39,N,1.0,74,42,1.0,0.74,7.2,...,0.5,1.94,0.0,NaN,1.0,11.64,1.0,1.0,0.00,0.0
1,2,2025-11-01 00:18:52,2025-11-01 00:24:27,N,1.0,74,42,2.0,0.95,7.2,...,0.5,0.00,0.0,NaN,1.0,9.70,2.0,1.0,0.00,0.0
2,2,2025-11-01 01:03:14,2025-11-01 01:15:24,N,1.0,83,160,1.0,2.19,13.5,...,0.5,5.00,0.0,NaN,1.0,21.00,1.0,1.0,0.00,0.0
3,2,2025-11-01 00:10:57,2025-11-01 00:24:53,N,1.0,166,127,1.0,5.44,24.7,...,0.5,0.50,0.0,NaN,1.0,27.70,1.0,1.0,0.00,0.0
4,1,2025-11-01 00:03:48,2025-11-01 00:19:38,N,1.0,166,262,1.0,3.20,18.4,...,1.5,1.00,0.0,NaN,1.0,24.65,1.0,1.0,2.75,0.0


In [6]:
df_zones.head()

,LocationID,Borough,Zone,service_zone
0,1,EWR,Newark Airport,EWR
1,2,Queens,Jamaica Bay,Boro Zone
2,3,Bronx,Allerton/Pelham Gardens,Boro Zone
3,4,Manhattan,Alphabet City,Yellow Zone
4,5,Staten Island,Arden Heights,Boro Zone


In [7]:
df_trips.dtypes

VendorID                          int32
lpep_pickup_datetime     datetime64[us]
lpep_dropoff_datetime    datetime64[us]
store_and_fwd_flag               object
RatecodeID                      float64
PULocationID                      int32
DOLocationID                      int32
passenger_count                 float64
trip_distance                   float64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
ehail_fee                       float64
improvement_surcharge           float64
total_amount                    float64
payment_type                    float64
trip_type                       float64
congestion_surcharge            float64
cbd_congestion_fee              float64
dtype: object

In [8]:
df_trips.shape

(46912, 21)

In [9]:
df_zones.dtypes

LocationID       int64
Borough         object
Zone            object
service_zone    object
dtype: object

In [10]:
df_zones.shape

(265, 4)

# Creating SQL Schema

In [11]:
print(pd.io.sql.get_schema(df_trips, name='green_trips', con=engine))


CREATE TABLE green_trips (
	"VendorID" INTEGER, 
	lpep_pickup_datetime TIMESTAMP WITHOUT TIME ZONE, 
	lpep_dropoff_datetime TIMESTAMP WITHOUT TIME ZONE, 
	store_and_fwd_flag TEXT, 
	"RatecodeID" FLOAT(53), 
	"PULocationID" INTEGER, 
	"DOLocationID" INTEGER, 
	passenger_count FLOAT(53), 
	trip_distance FLOAT(53), 
	fare_amount FLOAT(53), 
	extra FLOAT(53), 
	mta_tax FLOAT(53), 
	tip_amount FLOAT(53), 
	tolls_amount FLOAT(53), 
	ehail_fee FLOAT(53), 
	improvement_surcharge FLOAT(53), 
	total_amount FLOAT(53), 
	payment_type FLOAT(53), 
	trip_type FLOAT(53), 
	congestion_surcharge FLOAT(53), 
	cbd_congestion_fee FLOAT(53)
)




In [12]:
print(pd.io.sql.get_schema(df_zones, name='zones', con=engine))


CREATE TABLE zones (
	"LocationID" BIGINT, 
	"Borough" TEXT, 
	"Zone" TEXT, 
	service_zone TEXT
)




Creating Tables without any row data

In [13]:
df_trips.head(0).to_sql(name='green_trips', con=engine, if_exists='replace')
df_zones.head(0).to_sql(name='zones', con=engine, if_exists='replace')

0

# Inserting Rows to SQL Tables with specified chunksize

In [14]:
len(df_trips)

46912

In [15]:
len(df_zones)

265

For `df_trips`, there are 46912 rows to be inserted. The number of rows is not that many and we have sufficient memory to place all the rows into a buffer but for the sake of learning, I will use chunksize to demonstrate scenarios when the number of rows is significant.

In [16]:
# Reading parquet files
pf = pq.ParquetFile(trips_parquet_file_loc)

# Processing parquet file in iterations (chunksize of 10_000)
for batch in tqdm(pf.iter_batches(batch_size=10_000)):
    df_chunk = batch.to_pandas()
    df_chunk.to_sql(
        "green_trips",
        engine,
        if_exists="append",
        index=False,
        method="multi"
    )

0it [00:00, ?it/s]

For `df_zones`, only 265 rows so we are good to perform a bulk insert without chunksize.

In [17]:
df_zones.to_sql(name='zones', con=engine, if_exists='append')

265